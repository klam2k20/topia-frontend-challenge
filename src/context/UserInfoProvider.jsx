import React, { createContext, useContext, useState } from 'react';

export const UserInfoContext = createContext();

export const UserInfoProvider = ({ children }) => {
  const [position, setPosition] = useState({ x: 800, y: 400 });
  const [screenSize, setScreenSize] = useState({
    width: window.innerWidth,
    height: window.innerHeight,
  });

  const updateUserPosition = (p) => setPosition(p);
  const updateScreenSize = (s) => setScreenSize(s);

  return (
    <UserInfoContext.Provider
      value={{
        position,
        screenSize,
        updateUserPosition,
        updateScreenSize,
      }}>
      {children}
    </UserInfoContext.Provider>
  );
};

export const useUserInfo = () => useContext(UserInfoContext);
