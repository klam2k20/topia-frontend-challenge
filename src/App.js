import { Container } from '@material-ui/core';
import { Stack } from '@mui/material';
import { useEffect, useState } from 'react';
import UpdateViewModal from './components/UpdateViewModal';
import UserInfo from './components/UserInfo';
import { UserList } from './components/UserList';
import { useUserInfo } from './context/UserInfoProvider';
import './styles/App.css';
import listUsersInView from './utils/listUsersInView';

function App() {
  const [visibleUsers, setVisibleUsers] = useState([]);
  const { position, screenSize } = useUserInfo();

  useEffect(() => {
    const users = listUsersInView(position.x, position.y, screenSize.width, screenSize.height);
    setVisibleUsers(users);
  }, []);

  return (
    <Container maxWidth='md' className='App'>
      <Stack direction={'column'} spacing={4}>
        <UserInfo />
        <UserList users={visibleUsers} />
        <UpdateViewModal updateVisibleUsers={(users) => setVisibleUsers(users)} />
      </Stack>
    </Container>
  );
}

export default App;
