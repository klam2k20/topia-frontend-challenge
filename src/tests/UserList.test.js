import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import { UserInfoProvider } from '../context/UserInfoProvider';
import App from '../App';

describe('UserList', () => {
  it('renders correctly', () => {
    render(
      <UserInfoProvider>
        <App />
      </UserInfoProvider>
    );

    const userlist = screen.getByLabelText("List of all peers within the user's viewport");
    expect(userlist).toMatchSnapshot();
  });

  it('Empty UserList renders correctly', async () => {
    render(
      <UserInfoProvider>
        <App />
      </UserInfoProvider>
    );

    const openModal = screen.getByText('Update User View');
    fireEvent.click(openModal);

    await screen.findByText('Update View');

    const screenWidthInput = screen.getByRole('spinbutton', { name: /Screen Width/i });
    const screenHeightInput = screen.getByRole('spinbutton', { name: /Screen Height/i });
    fireEvent.change(screenWidthInput, { target: { value: '100' } });
    fireEvent.change(screenHeightInput, { target: { value: '100' } });

    const update = screen.getByText('Update');
    fireEvent.click(update);

    await screen.findByText('No Peers Around');

    const icon = screen.getByLabelText('No Peers Around Icon');
    const label = screen.getByText('No Peers Around');

    expect(icon).toBeInTheDocument();
    expect(label).toBeInTheDocument();
  });

  it("Updating the user's position and screen dimensions via the UpdateViewModal updates UserList", async () => {
    render(
      <UserInfoProvider>
        <App />
      </UserInfoProvider>
    );

    const title = screen.getByText("User's Information");
    const positionLabel = screen.getByText('Position:');
    const screenSizeLabel = screen.getByText('Screen Size:');
    const initialPosition = screen.getByText('(800, 400)');
    const initialScreenSize = screen.getByText(`${window.innerWidth} x ${window.innerHeight}`);

    expect(title).toBeInTheDocument();
    expect(positionLabel).toBeInTheDocument();
    expect(screenSizeLabel).toBeInTheDocument();
    expect(initialPosition).toBeInTheDocument();
    expect(initialScreenSize).toBeInTheDocument();

    const openModal = screen.getByText('Update User View');
    fireEvent.click(openModal);

    await screen.findByText('Update View');

    const xPosition = screen.getByRole('spinbutton', { name: /X Position/i });
    const yPosition = screen.getByRole('spinbutton', { name: /Y Position/i });
    const screenWidthInput = screen.getByRole('spinbutton', { name: /Screen Width/i });
    const screenHeightInput = screen.getByRole('spinbutton', { name: /Screen Height/i });
    const updateBtn = screen.getByRole('button', { name: /Update/i });

    fireEvent.change(xPosition, { target: { value: '0' } });
    fireEvent.change(yPosition, { target: { value: '0' } });
    fireEvent.change(screenWidthInput, { target: { value: '100' } });
    fireEvent.change(screenHeightInput, { target: { value: '250' } });
    fireEvent.click(updateBtn);

    await screen.findByText('(0, 0)');

    const idLabel = screen.getByText('ID');
    const usernameLabel = screen.getByText('Username');
    const distanceLabel = screen.getByText('Distance (px)');
    const broadcasterLabel = screen.getByText('Is Broadcaster');
    const username = screen.getByText('user2');

    expect(idLabel).toBeInTheDocument();
    expect(usernameLabel).toBeInTheDocument();
    expect(distanceLabel).toBeInTheDocument();
    expect(broadcasterLabel).toBeInTheDocument();
    expect(username).toBeInTheDocument();
  });

  // it("UserList only renders peer avatars whose avatars are fully within the user's viewport", async () => {
  //   render(
  //     <UserInfoProvider>
  //       <App />
  //     </UserInfoProvider>
  //   );

  //   const openModal = screen.getByText('Update User View');
  //   fireEvent.click(openModal);

  //   await screen.findByText('Update View');

  //   const xPosition = screen.getByRole('spinbutton', { name: /X Position/i });
  //   const yPosition = screen.getByRole('spinbutton', { name: /Y Position/i });
  //   const screenWidthInput = screen.getByRole('spinbutton', { name: /Screen Width/i });
  //   const screenHeightInput = screen.getByRole('spinbutton', { name: /Screen Height/i });
  //   const updateBtn = screen.getByRole('button', { name: /Update/i });

  //   fireEvent.change(xPosition, { target: { value: '0' } });
  //   fireEvent.change(yPosition, { target: { value: '0' } });
  //   fireEvent.change(screenWidthInput, { target: { value: '100' } });
  //   fireEvent.change(screenHeightInput, { target: { value: '100' } });
  //   fireEvent.click(updateBtn);

  //   await screen.findByText('No Peers Around');

  //   const icon = screen.getByLabelText('No Peers Around Icon');
  //   const label = screen.getByText('No Peers Around');

  //   expect(icon).toBeInTheDocument();
  //   expect(label).toBeInTheDocument();
  // });

  it("UserList renders peer avatars in sorted order based on distance from user's position", async () => {
    render(
      <UserInfoProvider>
        <App />
      </UserInfoProvider>
    );

    const openModal = screen.getByText('Update User View');
    fireEvent.click(openModal);

    await screen.findByText('Update View');

    const xPosition = screen.getByRole('spinbutton', { name: /X Position/i });
    const yPosition = screen.getByRole('spinbutton', { name: /Y Position/i });
    const screenWidthInput = screen.getByRole('spinbutton', { name: /Screen Width/i });
    const screenHeightInput = screen.getByRole('spinbutton', { name: /Screen Height/i });
    const updateBtn = screen.getByRole('button', { name: /Update/i });

    fireEvent.change(xPosition, { target: { value: '0' } });
    fireEvent.change(yPosition, { target: { value: '0' } });
    fireEvent.change(screenWidthInput, { target: { value: '2000' } });
    fireEvent.change(screenHeightInput, { target: { value: '2000' } });
    fireEvent.click(updateBtn);

    await screen.findByText('(0, 0)');

    const distanceCells = screen.getAllByLabelText('distance');

    const distanceValues = distanceCells.map((cell) => {
      // eslint-disable-next-line testing-library/no-node-access
      const pElement = cell.querySelector('p');
      return pElement.textContent.trim();
    });

    for (let i = 0; i < distanceValues.length - 1; i++) {
      expect(+distanceValues[i]).toBeLessThanOrEqual(+distanceValues[i + 1]);
    }
  });
});
