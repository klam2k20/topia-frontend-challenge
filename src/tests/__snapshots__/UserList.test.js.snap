// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserList renders correctly 1`] = `
<div
  aria-label="List of all peers within the user's view"
  class="MuiStack-root css-jfdv4h-MuiStack-root"
>
  <h6
    class="MuiTypography-root MuiTypography-h6 MuiTypography-alignLeft"
  >
    Visible Users
  </h6>
  <div
    class="MuiTableContainer-root"
  >
    <table
      aria-label="Visible peers within the user screen"
      class="MuiTable-root"
    >
      <thead
        class="MuiTableHead-root"
        style="background-color: black;"
      >
        <tr
          class="MuiTableRow-root MuiTableRow-head"
        >
          <th
            class="MuiTableCell-root MuiTableCell-head"
            scope="col"
          >
            <p
              class="MuiTypography-root white-text MuiTypography-body1"
            >
              ID
            </p>
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter"
            scope="col"
          >
            <p
              class="MuiTypography-root white-text MuiTypography-body1"
            >
              Username
            </p>
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter"
            scope="col"
          >
            <p
              class="MuiTypography-root white-text MuiTypography-body1"
            >
              Distance (px)
            </p>
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter"
            scope="col"
          >
            <p
              class="MuiTypography-root white-text MuiTypography-body1"
            >
              Is Broadcaster
            </p>
          </th>
        </tr>
      </thead>
      <tbody
        class="MuiTableBody-root"
      >
        <tr
          aria-label="user8 is 279 from user positon"
          class="MuiTableRow-root"
        >
          <td
            class="MuiTableCell-root MuiTableCell-body"
          >
            <p
              class="MuiTypography-root MuiTypography-body1"
            >
              8
            </p>
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter"
          >
            <p
              class="MuiTypography-root MuiTypography-body1"
            >
              user8
            </p>
          </td>
          <td
            aria-label="distance"
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter"
          >
            <p
              class="MuiTypography-root MuiTypography-body1"
            >
               
              279
            </p>
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter"
          >
            <p
              class="MuiTypography-root MuiTypography-body1"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
                data-testid="CellTowerIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="m7.3 14.7 1.2-1.2c-1-1-1.5-2.3-1.5-3.5 0-1.3.5-2.6 1.5-3.5L7.3 5.3c-1.3 1.3-2 3-2 4.7s.7 3.4 2 4.7zM19.1 2.9l-1.2 1.2c1.6 1.6 2.4 3.8 2.4 5.9 0 2.1-.8 4.3-2.4 5.9l1.2 1.2c2-2 2.9-4.5 2.9-7.1 0-2.6-1-5.1-2.9-7.1z"
                />
                <path
                  d="M6.1 4.1 4.9 2.9C3 4.9 2 7.4 2 10c0 2.6 1 5.1 2.9 7.1l1.2-1.2c-1.6-1.6-2.4-3.8-2.4-5.9 0-2.1.8-4.3 2.4-5.9zm10.6 10.6c1.3-1.3 2-3 2-4.7-.1-1.7-.7-3.4-2-4.7l-1.2 1.2c1 1 1.5 2.3 1.5 3.5 0 1.3-.5 2.6-1.5 3.5l1.2 1.2zM14.5 10c0-1.38-1.12-2.5-2.5-2.5S9.5 8.62 9.5 10c0 .76.34 1.42.87 1.88L7 22h2l.67-2h4.67l.66 2h2l-3.37-10.12c.53-.46.87-1.12.87-1.88zm-4.17 8L12 13l1.67 5h-3.34z"
                />
              </svg>
            </p>
          </td>
        </tr>
        <tr
          aria-label="user9 is 279 from user positon"
          class="MuiTableRow-root"
        >
          <td
            class="MuiTableCell-root MuiTableCell-body"
          >
            <p
              class="MuiTypography-root MuiTypography-body1"
            >
              9
            </p>
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter"
          >
            <p
              class="MuiTypography-root MuiTypography-body1"
            >
              user9
            </p>
          </td>
          <td
            aria-label="distance"
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter"
          >
            <p
              class="MuiTypography-root MuiTypography-body1"
            >
               
              279
            </p>
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter"
          >
            <p
              class="MuiTypography-root MuiTypography-body1"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
                data-testid="CellTowerIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="m7.3 14.7 1.2-1.2c-1-1-1.5-2.3-1.5-3.5 0-1.3.5-2.6 1.5-3.5L7.3 5.3c-1.3 1.3-2 3-2 4.7s.7 3.4 2 4.7zM19.1 2.9l-1.2 1.2c1.6 1.6 2.4 3.8 2.4 5.9 0 2.1-.8 4.3-2.4 5.9l1.2 1.2c2-2 2.9-4.5 2.9-7.1 0-2.6-1-5.1-2.9-7.1z"
                />
                <path
                  d="M6.1 4.1 4.9 2.9C3 4.9 2 7.4 2 10c0 2.6 1 5.1 2.9 7.1l1.2-1.2c-1.6-1.6-2.4-3.8-2.4-5.9 0-2.1.8-4.3 2.4-5.9zm10.6 10.6c1.3-1.3 2-3 2-4.7-.1-1.7-.7-3.4-2-4.7l-1.2 1.2c1 1 1.5 2.3 1.5 3.5 0 1.3-.5 2.6-1.5 3.5l1.2 1.2zM14.5 10c0-1.38-1.12-2.5-2.5-2.5S9.5 8.62 9.5 10c0 .76.34 1.42.87 1.88L7 22h2l.67-2h4.67l.66 2h2l-3.37-10.12c.53-.46.87-1.12.87-1.88zm-4.17 8L12 13l1.67 5h-3.34z"
                />
              </svg>
            </p>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
